### PR TITLE
Added a user model

### PR DESCRIPTION
**Pull request by mattdharmon**
As well as test with factories, added a bcrypt lib for password hashing, and a nanobox (boxfile.yml) for consistent dev env.

### CODE REVIEW

File: config/dev.exs [hunk 1]
```diff
@@ -39,8 +39,8 @@ config :phoenix, :stacktrace_depth, 20
 # Configure your database
 config :real_world, RealWorld.Repo,
   adapter: Ecto.Adapters.Postgres,
-  username: "postgres",
-  password: "postgres",
-  database: "real_world_dev",
-  hostname: "localhost",
+  username: System.get_env("DATA_DB_USER"),
```
Comment (brwr) [inline config/dev.exs]: Just a heads up: I think we still have an open conversation about whether to use environment variables and what to call them. We may ask you guys to change this in the future. Or we may not. Like I said -- haven't come to an answer yet. 👍 😄

Comment (mattdharmon) [inline config/dev.exs]: Understandable.

**Comment by brwr** [comment]
Nice PR! Thanks for taking the time!

I don't know anything about Elixir, but hopefully @BrainMaestro will stop by later and give us some thoughts.

**Comment by mattdharmon** [comment]
I'm fairly new to Elixir myself (a few months of personal projects) but when I saw this project, I thought why not help out with what I learned already.

My personal project is already set up with a JWT token and the what not, but I thought I put that in a different pull request after the user one get's accepted.

File: boxfile.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  before_live:
+    web.main:
+      - mix ecto.create --quiet
+      - mix ecto.migrate
+
+data.db:
+  image: nanobox/postgresql:9.5
+
+web.main:
+  start: mix phx.server
```
Comment (sotojuan) [inline boxfile.yml]: What is this for? We should discuss with @BrainMaestro before adding any deployment set up. Sorry in advance if this was already discussed.

Comment (mattdharmon) [inline boxfile.yml]: Nanobox config file for local dev work.

**Comment by mattdharmon** [comment]
I removed the local dev env and restored the configs to be more like master so that this is a bit of a cleaner PR.

File: config/dev.exs [hunk 3]
```diff
@@ -41,6 +41,6 @@ config :real_world, RealWorld.Repo,
   adapter: Ecto.Adapters.Postgres,
   username: "postgres",
   password: "postgres",
-  database: "real_world_dev",
```
Comment (BrainMaestro) [inline config/dev.exs]: Any reason why this changed?

Comment (sotojuan) [inline config/dev.exs]: Because of the new name (`RealWorld` vs `Realworld`). Happened to me too.

Comment (BrainMaestro) [inline config/dev.exs]: Yeah, but the current master has it as `real_world_dev` now. So a rebase should set everything straight. Sorry about that.

Comment (mattdharmon) [inline config/dev.exs]: Okay, will rebase, one min.

**Comment by mattdharmon** [comment]
Okay, updated.

**Comment by BrainMaestro** [comment]
Thanks. This looks good

**Comment by BrainMaestro** [comment]
@mattdharmon, any chance we could split the `ex_machina` factory part into a separate PR? Sorry to ask, but I feel it would make this a lot clearer. Especially for people that are not so familiar with it. 

Also, it was throwing an error `Could not start application ex_machina: could not find application file: ex_machina.app`

**Comment by mattdharmon** [comment]
@BrainMaestro I can do that, but will take a bit. Did you do a deps.get because I have it working on my machine.

**Comment by BrainMaestro** [comment]
@mattdharmon that's fine, take your time.

File: lib/users/users.ex [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  @doc """
+  Binds a map (%{}) to a User struct (%User{}) and
+  validates the struct to a set of requirements.
+  """
+  defp user_changeset(%User{} = user, attrs) do
+    user
+    |> cast(attrs, [:username, :password, :email])
+    |> validate_required([:username, :password])
```
Comment (joseluistorres) [inline lib/users/users.ex]: @mattdharmon `:email` is also required https://github.com/gothinkster/realworld/tree/master/api#registration

File: mix.exs [hunk 5]
```diff
@@ -17,11 +17,11 @@ defmodule RealWorld.Mixfile do
   # Type `mix help compile.app` for more information.
   def application do
     [mod: {RealWorld.Application, []},
-     extra_applications: [:logger, :runtime_tools]]
+     extra_applications: [:logger, :runtime_tools, :comeonin, :ex_machina]]
```
Comment (supernova23) [inline mix.exs]: you are requesting the ex_machina dependency just for the test environment
~~~~
{:ex_machina, "~> 2.0", only: :test}]
~~~~
so we need to load it just when starting the app in the test environments
~~~~
extra_applications: applications(Mix.env)

def applications(:test), do: [:logger, :runtime_tools, :comeonin, :ex_machina]
def applications(_), do: [:logger, :runtime_tools, :comeonin]
~~~~

Comment (mattdharmon) [inline mix.exs]: Yeah, but @BrainMaestro thought it would be better to remove it for now and make a new PR for it.

**Comment by mattdharmon** [comment]
Hmm, I pushed in new updates that also fixed the conflicts, but is seems to not have updated yet.

**Comment by BrainMaestro** [comment]
Hey @mattdharmon. Thanks for the PR. This looks great! 
One last thing, could you move the `users` folder to the `real_world` one? So that it'll match the new structure for `1.3` and also be similar to `blog`

**Comment by mattdharmon** [comment]
Sure thing, min.

**Comment by BrainMaestro** [comment]
Awesome. You are welcome to pick up something else.

**Pull request merged**